pkg/cachedir/manager.go

```diff
@@ -0,0 +1,186 @@
+}
+
+type HashedKey string
+
+// hashKey hashes the key to create the directory name because otherwise tche path
+// would be too long. This hash needs to be secure or else multiple downloads could
+// share the same cache directory and cause unexpected and potentially malicious behaviour.
+func hashKey(key string) HashedKey {
+	hash := sha256.New()
+	hash.Write([]byte(key))
+	encoded := make([]byte, hash.Size()*2)
+	hex.Encode(encoded, hash.Sum(nil))
+	return HashedKey(encoded)
+}
+
+func (c *manager) cachePathForKeyHash(keyHash HashedKey) string {
+	return filepath.Join(c.cacheRoot, string(keyHash))
+}
+
+func dirSize(path string) (uint64, error) {
+	var size uint64
+	err := filepath.Walk(path, func(_ string, info os.FileInfo, err error) error {
+		if info != nil && !info.IsDir() {
+			size += uint64(info.Size())
+		}
+		return err
+	})
+	return size, err
+}
+
+func (c *manager) sizeEstimateByKeyHash(keyHash HashedKey) uint64 {
+	dir := c.cachePathForKeyHash(keyHash)
+	size, err := dirSize(dir)
+	if err != nil {
+		logrus.Warnf("failed to determine size of directory %s: %s", dir, err)
+		return 0
+	}
+	return size
+}
+
+func (c *manager) cleanUpByKeyHash(keyHash HashedKey) error {
+	return os.RemoveAll(c.cachePathForKeyHash(keyHash))
+}
+
+func (c *manager) GetDir(key string) string {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	logrus.Debugf("cachedir manager get: %s", key)
+	hashedKey := hashKey(key)
+	if _, ok := c.inUse[hashedKey]; ok {
+		panic(fmt.Errorf("tried to get a cachedir key (%s) twice without releasing/deleting", key))
+	}
+	c.inUse[hashedKey] = struct{}{}
+
+	return c.cachePathForKeyHash(hashedKey)
+}
+
+func (c *manager) ReleaseDir(key string) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	logrus.Debugf("cachedir manager release: %s", key)
+	hashedKey := hashKey(key)
+	if _, ok := c.inUse[hashedKey]; !ok {
+		panic(fmt.Errorf("tried to release a cachedir key (%s) without getting it first", key))
+	}
+	delete(c.inUse, hashedKey)
+}
+
+func (c *manager) DeleteDir(key string) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	logrus.Debugf("cachedir manager delete: %s", key)
+	hashedKey := hashKey(key)
+	if _, ok := c.inUse[hashedKey]; !ok {
+		panic(fmt.Errorf("tried to delete a cachedir key (%s) without getting it first", key))
+	}
+	delete(c.inUse, hashedKey)
+	return c.cleanUpByKeyHash(hashedKey)
+}
+
+func (c *manager) CollectGarbage() (uint64, error) {
+	// safety check to make sure we don't try to delete files in / or pwd
+	if c.cacheRoot == "" {
+		return 0, nil
+	}
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	// mark
+	allKeyHashes := map[HashedKey]struct{}{}
+	cacheDirs, err := ioutil.ReadDir(c.cacheRoot)
+	if err != nil {
+		logrus.Warnf("Failed to list %s: %s", c.cacheRoot, err)
+	}
+	for _, cacheDir := range cacheDirs {
+		cacheDirName := cacheDir.Name()
+		keyHash := HashedKey(filepath.Base(cacheDirName))
+
+		// safety check to make sure we don't delete anything that's not a hash
+		// we created
+		decoded := make([]byte, sha256.Size)
```

Comment:
`var decoded [sha256.Size]byte`

Then:

```
hex.Decode(decoded[:], ...)
```

This preserves the possibility of keeping this stack allocated. Maybe not today, but a better compiler will catch that.